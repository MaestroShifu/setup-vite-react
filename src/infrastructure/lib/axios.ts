import axios, { AxiosRequestConfig } from 'axios';

const headers = {
    'Content-Type': 'application/json'
};

type AxiosDataParams = Record<string, string | number | boolean>;

const httpAxios = (baseURL: string) => {
    const http = axios.create({
        baseURL,
        headers
    });

    return {        
        get: async <T>(path: string, params?: AxiosDataParams, config?: AxiosRequestConfig) => {
            const response = await http.get<T>(path, { ...config, params });
            return response.data;
        },
        post: async <T>(path: string, data?: AxiosDataParams, config?: AxiosRequestConfig) => {
            const response = await http.post<T>(path, data, { ...config });
            return response.data;
        },
        put: async <T>(path: string, data?: AxiosDataParams, config?: AxiosRequestConfig) => {
            const response = await http.put<T>(path, data, { ...config });
            return response.data;
        },
        delete: async <T>(path: string, params?: AxiosDataParams, config?: AxiosRequestConfig) => {
            const response = await http.delete<T>(path, { ...config, params });
            return response.data;
        }
    }
};

export type HttpAxios = ReturnType<typeof httpAxios>;

export default httpAxios;