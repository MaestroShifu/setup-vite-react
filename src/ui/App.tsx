import { useState, useEffect } from 'react';
import logoURL from './logo.svg';
import './App.css';
import { Product } from '../domain/entities/product';
import productService from '../infrastructure/services/product.service';

function App() {
  const [count, setCount] = useState(0);
  const [products, setProducts] = useState<Product[]>([]);

  useEffect(() => {
    const loadProducts = async () => {
      const product = await productService.getProducts();
      setProducts(product);
    };

    loadProducts();
  }, []);

  return (
    <div className="App">
      <header className="App-header">
        <img
          src={logoURL}
          data-testid="img-logo"
          className="App-logo"
          alt="logo"
        />
        <p data-testid="title-app">Hello Vite + React!</p>
        <p>
          <button
            data-testid="button-count"
            type="button"
            onClick={() => setCount((c) => c + 1)}
          >
            count is: {count}
          </button>
        </p>
        <ol>
          {products.map((product) => (<li key={product.name}>{product.name}</li>))}  
        </ol>
      </header>
    </div>
  );
}

export default App;
