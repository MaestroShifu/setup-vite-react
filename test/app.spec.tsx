import React from 'react';
import { cleanup, render, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import axios from 'axios';
import App from '../src/ui/App';

jest.mock('axios');
const mockedAxios = jest.mocked(axios, true);

afterEach(cleanup); // Tenemos que limpiar siempre nuestro DOM Virtual

// Guia de testing library https://medium.com/swlh/beginners-guide-to-react-testing-with-testing-library-97f23b4af40f
describe('<App component>', () => {
  beforeEach(() => {
    mockedAxios.create.mockImplementation(() => axios);
    mockedAxios.get.mockResolvedValue({
      data: [
        {
          id: '1648690803050',
          name: 'marta',
          description: '',
          imageUrl: '',
          price: 0
        }
      ]
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('Snapshot of Body', async () => {
    const { asFragment } = render(<App />);
    await waitFor(() => {
      expect(asFragment()).toMatchSnapshot();
      expect(mockedAxios.get).toBeCalledTimes(1);
    });
  });

  it('finding title app', async () => {
    const { getByText } = render(<App />);
    await waitFor(() => {
      expect(getByText('Hello Vite + React!')).toBeInTheDocument(); // -> Documentacion jest-dom https://github.com/testing-library/jest-dom
      expect(mockedAxios.get).toBeCalledTimes(1);
    });
  });

  it('finding title with TestId', async () => {
    const { getByTestId } = render(<App />);
    await waitFor(() => {
      expect(getByTestId('title-app')).toHaveTextContent('Hello Vite + React!');
      expect(getByTestId('img-logo')).toBeInTheDocument();
      expect(mockedAxios.get).toBeCalledTimes(1);
    });
  });

  it('check Counter initial value', async () => {
    const { getByTestId } = render(<App />);
    await waitFor(() => {
      expect(getByTestId('button-count')).toHaveTextContent('0');
      expect(mockedAxios.get).toBeCalledTimes(1);
    });
  });

  it('check Counter value after a tap', async () => {
    const { getByTestId } = render(<App />);
    const button = getByTestId('button-count');
    userEvent.click(button);
    await waitFor(() => {
      expect(getByTestId('button-count')).toHaveTextContent('1');
      expect(mockedAxios.get).toBeCalledTimes(1);
    });
  });

  it('check Counter value after tapping twice', async () => {
    const { getByTestId } = render(<App />);
    const button = getByTestId('button-count');
    userEvent.click(button);
    userEvent.click(button);
    await waitFor(() => {
      expect(getByTestId('button-count')).toHaveTextContent('2');
      expect(mockedAxios.get).toBeCalledTimes(1);
    });
  });
});
