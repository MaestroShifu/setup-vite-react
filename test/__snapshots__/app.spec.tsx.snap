// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App component> Snapshot of Body 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <header
      class="App-header"
    >
      <img
        alt="logo"
        class="App-logo"
        data-testid="img-logo"
        src="SvgrURL"
      />
      <p
        data-testid="title-app"
      >
        Hello Vite + React!
      </p>
      <p>
        <button
          data-testid="button-count"
          type="button"
        >
          count is: 0
        </button>
      </p>
      <ol />
    </header>
  </div>
</DocumentFragment>
`;
